build: Link time garbage collection

fanquake opened the pull request:
Closes #18579.

Adds the linker flags required to turn on link time garbage collection
to the Linux gitian descriptors and Guix build. After some experimentation
I'm not convinced this is something that we should do for Windows builds
at this stage.

Size comparison for a gitian build of master (df303ceb650521dc7b1ba91e0eea383c387a5860) vs this PR (8c31ef9ec68549c4105e8115f74f435c71b3901c):

| binary | master (df303ceb650521dc7b1ba91e0eea383c387a5860) | PR (8c31ef9ec68549c4105e8115f74f435c71b3901c) | saving | 
| - | - | - | - |
| bitcoind | 10525336 | 9775672 | 7% |
| bitcoin-cli | 2055800 | 1113624 | 45% |
| bitcoin-qt | 31896456 | 31114024 | 2.5%
| bitcoin-tx | 2376152 | 1372536 | 42% |
| bitcoin-wallet | 5899168 | 3199808 | 45% |
| test_bitcoin | 17063328 | 15354784 | 10% |

fanquake committed:
build: perform link time garbage collection in gitian

Adds the linker flag required to turn on link time garbage collection
to the linux gitian descriptors. Combined with `-ffunction-sections` and
`-fdata-sections` this results in a between 5% and 45% reduction in file
size for the binaries we are shipping.

laanwj commented:
Concept ACK.

fanquake committed:
build: perform link time garbage collection in Guix

Adds the linker flags required to turn on link time garbage collection
to the Linux Guix build. Combined with `-ffunction-sections` and
`-fdata-sections` this results in a between ~5% and 45% reduction in file
size for the binaries we are shipping.

fanquake commented:
Rebased for #17929 and added a commit for Guix.

jonasschnelli commented:
Nice. Concept ACK.
Is there a clang equivalent for `--gc-sections`?
`fdata-sections` AFAIK is also supported in clang.
Could the clang linker flag be `-dead_strip`?

laanwj commented on contrib/gitian-descriptors/gitian-linux.yml:
```diff
@@ -44,9 +44,9 @@ script: |
-  HOST_CFLAGS="-O2 -g"
-  HOST_CXXFLAGS="-O2 -g"
-  HOST_LDFLAGS_BASE="-static-libstdc++ -Wl,-O2"
+  HOST_CFLAGS="-O2 -g -ffunction-sections -fdata-sections"
```
These flags are effective for the depends build too, I hope?

fanquake commented:
I've been doing some experiments with Windows builds, and passing `--gc-sections` there currently results in an _increase_ in binary size. I still need to test using a more recent version of binutils.

> Is there a clang equivalent for --gc-sections?

`gc-sections` is a linker flag, so as long as Clang forwards it through, and whichever linker you are using understands it, it should work. Note that ld64 on macOS does not.

> fdata-sections AFAIK is also supported in clang.

Clang supports `-ffunction-sections` and  `-fdata-sections`. However only for ELF targets.
```bash
-fdata-sections         Place each data in its own section (ELF Only)
-ffunction-sections     Place each function in its own section (ELF Only)
```

> Could the clang linker flag be -dead_strip?

We do currently pass `dead_strip` to `ld64` for our macOS builds; and see size reductions. For example, compiling master with and without:

| Binary | dead_strip | no dead_strip |
| - | - | - |
| bitcoind | 10900836 | 11741428 |
| bitcoin-qt | 28593032 | 32178676 |
| bitcoin-cli | 466896 | 1065220 |
| bitcoin-tx | 1196536 | 1671900 |
| bitcoin-wallet | 3054504 | 5581956 |

fanquake commented:
I've updated this to drop the changes for Windows for now. After some more experimenting it wasn't clear that it was working as we'd like and I was still seeing increased binary sizes with newer binutils.

I have also updated the PR description with a comparison of the size differences for a gitian build of master vs this branch.

practicalswift commented:
Concept ACK

fanquake commented:
I rebased this and ran some gitian builds to compare master (816132e6eb23c06287c1e820cc0990d2317c2f10) and this PR (d05d56047a38686c70dc066d010ed8f06990508b) (also included 0.20.1):
```bash
# ls -la bitcoin-0.20.1/bin/
2055800 Jul 31 12:01 bitcoin-cli
31843176 Jul 31 12:01 bitcoin-qt
2376152 Jul 31 12:01 bitcoin-tx
5952832 Jul 31 12:01 bitcoin-wallet
10312344 Jul 31 12:01 bitcoind
16702880 Jul 31 12:01 test_bitcoin

# ls -la bitcoin-816132e6eb23/bin/
4689528 Nov 21 06:47 bitcoin-cli
36844872 Nov 21 06:47 bitcoin-qt
4792088 Nov 21 06:47 bitcoin-tx
10000192 Nov 21 06:47 bitcoin-wallet
15277080 Nov 21 06:47 bitcoind
21103040 Nov 21 06:47 test_bitcoin

# ls -la bitcoin-d05d56047a38/bin
4439576 Apr 15  2020 bitcoin-cli
34407752 Apr 15  2020 bitcoin-qt
4505304 Apr 15  2020 bitcoin-tx
7214880 Apr 15  2020 bitcoin-wallet
14953496 Apr 15  2020 bitcoind
20516928 Apr 15  2020 test_bitcoin
```

All of our shipped binaries have certainly grown a bit since 0.20.1. There's still a few mb to be saved in some cases.

dergoegge commented:
I confirmed the space savings by running the gitian builds on master and this PR. I also read up on how the used options (`functions-sections`, `data-sections`, `gc-sections`) work and if there any unwanted side effects.

Two questions:
1. *AFAICT the options are not active in the depends build. Why not?* 

https://github.com/bitcoin/bitcoin/blob/d05d56047a38686c70dc066d010ed8f06990508b/contrib/gitian-descriptors/gitian-linux.yml#L96-L99

2. *The gcc docs warn to only use these options if there is a significant benefit from doing so because certain compiler optimizations wont work while they are enabled. Is it ok to ignore that warning? If so, why?*

From https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html:
> Only use these options when there are significant benefits from doing so. [...] **They prevent optimizations by the compiler and assembler** using relative locations inside a translation unit since the locations are unknown until link time. An example of such an optimization is relaxing calls to short call instructions.
